duration.go

```diff
@@ -22,28 +22,38 @@ func (dur Duration) MarshalText() ([]byte, error) {
 	return []byte(fmt.Sprintf("%02d:%02d:%02d.%03d", h, m, s, ms)), nil
 }
 
-// UnmarshalText implements the encoding.TextUnmarshaler interface.
-func (dur *Duration) UnmarshalText(data []byte) (err error) {
+// DecodeDuration converts a timestamp into a Duration nanoseconds
+func DecodeDuration(data []byte) (Duration, error) {
 	parts := strings.SplitN(string(data), ":", 3)
 	if len(parts) != 3 {
-		return fmt.Errorf("invalid duration: %s", data)
+		return Duration(-1), fmt.Errorf("invalid duration: %s", data)
 	}
+	var dur = Duration(0)
 	if i := strings.IndexByte(parts[2], '.'); i > 0 {
 		ms, err := strconv.ParseInt(parts[2][i+1:], 10, 32)
 		if err != nil || ms < 0 || ms > 999 {
-			return fmt.Errorf("invalid duration: %s", data)
+			return Duration(-1), fmt.Errorf("invalid duration: %s", data)
 		}
 		parts[2] = parts[2][:i]
-		*dur += Duration(ms) * Duration(time.Millisecond)
+		dur += Duration(ms) * Duration(time.Millisecond)
 	}
 	f := Duration(time.Second)
 	for i := 2; i >= 0; i-- {
 		n, err := strconv.ParseInt(parts[i], 10, 32)
 		if err != nil || n < 0 || n > 59 {
-			return fmt.Errorf("invalid duration: %s", data)
+			return Duration(-1), fmt.Errorf("invalid duration: %s", data)
 		}
-		*dur += Duration(n) * f
+		dur += Duration(n) * f
 		f *= 60
 	}
+	return dur, nil
+
+}
+
+// UnmarshalText implements the encoding.TextUnmarshaler interface.
+func (dur *Duration) UnmarshalText(data []byte) (err error) {
+	// swallow the error since parsing a non-compliant duration will throw an error
+	d, _ := DecodeDuration(data)
```

Comment:
you should only swallow the error when the Duration is actually empty. So I suggest add a variable like `ErrDurationEmpty = errors.New("duration is empty")` and return that from `DecodeDuration`. Then you can do `if err == ErrDurationEmpty { // swallow error } else { return err }`